contracts/pcv/PCVDepositAggregator.sol

```diff
@@ -0,0 +1,444 @@
+        uint128 _bufferWeight
+    ) CoreRef(_core) {
+        require(_initialPCVDepositAddresses.length == _initialPCVDepositWeights.length, "Addresses and weights are not the same length!");
+        require(_assetManager != address(0x0), "Rewards asset manager cannot be null");
+        require(IRewardsAssetManager(_assetManager).getToken() == _token, "Rewards asset manager must be for the same token as this.");
+
+        assetManager = _assetManager;
+        token = _token;
+        bufferWeight = _bufferWeight;
+        totalWeight = bufferWeight;
+
+        _setContractAdminRole(keccak256("AGGREGATOR_ADMIN_ROLE"));
+
+        for (uint256 i=0; i < _initialPCVDepositAddresses.length; i++) {
+            require(IPCVDeposit(_initialPCVDepositAddresses[i]).balanceReportedIn() == _token, "Deposit token must be the same as for this aggregator.");
+            _addPCVDeposit(_initialPCVDepositAddresses[i], _initialPCVDepositWeights[i]);
+        }
+    }
+
+    // ---------- Public Functions -------------
+
+    /// @notice deposits tokens into sub-contracts (if needed)
+    /// @dev this is equivalent to half of a rebalance. the implementation is as follows:
+    /// 1. fill the buffer to maximum
+    /// 2. if buffer is full and there are still tokens unallocated, calculate the optimal istribution of tokens to sub-contracts
+    /// 3. distribute the tokens according the calcluations in step 2
+    function deposit() external override whenNotPaused {
+        // First grab the aggregator balance & the pcv deposit balances, and the sum of the pcv deposit balances
+        (uint256 actualAggregatorBalance, uint256 underlyingSum, uint256[] memory underlyingBalances) = _getUnderlyingBalancesAndSum();
+        uint256 totalBalance = underlyingSum + actualAggregatorBalance;
+
+        // Optimal aggregator balance is (bufferWeight/totalWeight) * totalBalance
+        uint256 optimalAggregatorBalance = bufferWeight * totalBalance / totalWeight;
+
+        // if actual aggregator balance is below optimal, we shouldn't deposit to underlying - just "fill up the buffer"
+        if (actualAggregatorBalance <= optimalAggregatorBalance) {
+            return;
+        }
+
+        // we should fill up the buffer before sending out to sub-deposits
+        uint256 amountAvailableForUnderlyingDeposits = optimalAggregatorBalance - actualAggregatorBalance;
+
+        // calculate the amount that each pcv deposit needs. if they have an overage this is 0.
+        uint256[] memory optimalUnderlyingBalances = _getOptimalUnderlyingBalances(totalBalance);
+        uint256[] memory amountsNeeded = optimalUnderlyingBalances.positiveDifference(underlyingBalances);
+        uint256 totalAmountNeeded = amountsNeeded.sum();
+
+        // calculate a scalar. this will determine how much we *actually* send to each underlying deposit.
+        Decimal.D256 memory scalar = Decimal.ratio(amountAvailableForUnderlyingDeposits, totalAmountNeeded);
```

Comment:
should probably assert this is <= 1 right?

Reply:
I don't believe it's possible to be > 1 but wouldn't hurt to add the assert.